### PR TITLE
yoav - Sysid Swerve Calibration

### PR DESCRIPTION
**Pull request by YHerm**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 1]
```diff
@@ -228,6 +248,20 @@ public void lockSwerve() {
         modules[3].setTargetState(frontLeftBackRight);
     }
 
+    /**
+     * Put swerve wheels in circle position, so it's ready to spin
+     */
+    public void readySpinSwerve() {
+        SwerveModuleState frontLeftBackRight = new SwerveModuleState(0, Rotation2d.fromDegrees(-45));
+        SwerveModuleState frontRightBackLeft = new SwerveModuleState(0, Rotation2d.fromDegrees(45));
```
Comment (hillelv) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:256]: magic
use eighth circle constant

File: src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java [hunk 2]
```diff
@@ -28,15 +30,46 @@ public class SwerveCommands {
 
     private static final Swerve SWERVE = RobotContainer.SWERVE;
 
+    private static final SysIdCalibrator STEER_CALIBRATOR = new SysIdCalibrator(
+            true,
+            SWERVE,
+            SWERVE::runModuleSteerByVoltage,
+            2
```
Comment (hillelv) [inline src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java:37]: magic

File: src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java [hunk 3]
```diff
@@ -28,15 +30,46 @@ public class SwerveCommands {
...
+    );
+
+    private static final SysIdCalibrator DRIVE_CALIBRATOR = new SysIdCalibrator(
+            true,
+            SWERVE,
+            SWERVE::runModulesDriveByVoltage,
+            2
```
Comment (hillelv) [inline src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java:44]: magic

File: src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java [hunk 4]
```diff
@@ -28,15 +30,46 @@ public class SwerveCommands {
...
-                SWERVE::getModulesDriveDistances,
-                SWERVE::getAbsoluteHeading,
-                SWERVE::runWheelRadiusCharacterization,
-                SWERVE::stop
+        Command command = new SequentialCommandGroup(
+                getReadySpinSwerveCommand(),
+                new WheelRadiusCharacterization(
+                        SWERVE,
+                        SwerveConstants.DRIVE_RADIUS_METERS,
+                        Rotation2d.fromRotations(0.5),
```
Comment (hillelv) [inline src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java:67]: half circle constant

File: src/main/java/frc/robot/subsystems/swerve/modules/Module.java [hunk 5]
```diff
@@ -93,11 +93,7 @@ public SwerveModuleState getTargetState() {
     }
 
     public boolean isAtTargetState() {
-        boolean isAtAngle = MathUtil.isNear(
-                MathUtil.angleModulus(getTargetState().angle.getRadians()),
-                MathUtil.angleModulus(getCurrentAngle().getRadians()),
-                ModuleConstants.ANGLE_TOLERANCE.getRadians()
-        );
+        boolean isAtAngle = isAtAngle(getTargetState().angle);
         boolean isAtVelocity = MathUtil.isNear(
```
Comment (hillelv) [inline src/main/java/frc/robot/subsystems/swerve/modules/Module.java:97]: for consistency move this to a different function

File: src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java (hunk 4 above)
Comment (YHerm) [reply src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java:67]: The number here 0.5 used for velocity and can be change and not for calculation so can’t use half circle but I can do another const

**Review by hillelv — APPROVED** [review]

**Review by YoniKiriaty — APPROVED** [review]

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 6]
```diff
@@ -213,15 +234,29 @@ public void pointWheels(MirrorableRotation2d targetAngle) {
...
     public void lockSwerve() {
-        SwerveModuleState frontLeftBackRight = new SwerveModuleState(0, Rotation2d.fromDegrees(45));
-        SwerveModuleState frontRightBackLeft = new SwerveModuleState(0, Rotation2d.fromDegrees(-45));
+        SwerveModuleState frontLeftBackRight = new SwerveModuleState(0, MathConstants.EIGHTH_CIRCLE);
+        SwerveModuleState frontRightBackLeft = new SwerveModuleState(0, MathConstants.EIGHTH_CIRCLE.unaryMinus());
+
+        modules[0].setTargetState(frontLeftBackRight);
+        modules[1].setTargetState(frontRightBackLeft);
+        modules[2].setTargetState(frontRightBackLeft);
+        modules[3].setTargetState(frontLeftBackRight);
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:243]: consider making this it's own function that takes in 4 states.

File: src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java [hunk 7]
```diff
@@ -54,6 +89,18 @@ public static Command getLockSwerveCommand() {
         return command;
     }
 
+    public static Command getReadySpinSwerveCommand() {
+        Command command = new FunctionalCommand(
+                () -> {},
+                SWERVE::readySpinSwerve,
+                inter -> {},
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java:96]: why is interpol in the code?
image

File: src/main/java/frc/robot/subsystems/swerve/Swerve.java [hunk 8]
```diff
@@ -213,15 +234,29 @@ public void pointWheels(MirrorableRotation2d targetAngle) {
...
+    }
+
+    /**
+     * Put swerve wheels in circle position, so it's ready to spin
+     */
+    public void readySpinSwerve() {
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/subsystems/swerve/Swerve.java:249]: did you mean `readyToSpinSwerve` or `readyToSpinSwerve`?

File: src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java [hunk 9]
```diff
@@ -46,19 +81,31 @@ public static Command getLockSwerveCommand() {
...
-                inter -> {},
+                interrupted -> {},
                 SWERVE::isModulesAtStates,
                 SWERVE
         );
         command.setName("Lock");
         return command;
     }
 
+    public static Command getReadySpinSwerveCommand() {
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/subsystems/swerve/SwerveCommands.java:92]: same as above

**Review by YoniKiriaty — APPROVED** [review]

**Pull request merged**